[Bug] [rllib] RNN sequencing is incorrect

### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.


### Ray Component

RLlib

### What happened + What you expected to happen

I would expect given two sequences `A, B`:
`[A, A, A, B, B]; seq_lens=[3, 2], obs.shape = [5, 1]`
would be padded to
`[A, A, A, B, B, *]; seq_lens=[3, 2], obs.shape = [2, 3, 1]`

This does not appear to be the case. For some reason rllib zero-pads `obs` to something besides `seq_lens.max()`. Even more worrisome is calling `torch.nonzero()` on the `input_dict`, which shows front-padded zeros to the observations. For example, printing `input_dict['obs'].reshape(B, T, -1) == 0` results in:

```
(PPO pid=74137)         [[ True,  True],
(PPO pid=74137)          [ True,  True],
(PPO pid=74137)          [ True,  True],
(PPO pid=74137)          [ True,  True],
(PPO pid=74137)          [ True,  True],
(PPO pid=74137)          [False, False],
(PPO pid=74137)          [False, False],
(PPO pid=74137)          [False, False],
(PPO pid=74137)          [False, False],
(PPO pid=74137)          [False, False],
(PPO pid=74137)          [False, False],
(PPO pid=74137)          [False, False],
(PPO pid=74137)          [False, False],
(PPO pid=74137)          [False, False],
(PPO pid=74137)          [False, False],
(PPO pid=74137)          [False, False],
(PPO pid=74137)          [False, False],
(PPO pid=74137)          [False, False],
(PPO pid=74137)          [False, False],
(PPO pid=74137)          [False, False],
(PPO pid=74137)          [False, False],
(PPO pid=74137)          [False, False],
(PPO pid=74137)          [False, False],
(PPO pid=74137)          [False, False],
(PPO pid=74137)          [False, False],
(PPO pid=74137)          [False, False],
(PPO pid=74137)          [ True,  True],
(PPO pid=74137)          [ True,  True],
(PPO pid=74137)          [ True,  True]],
```

The zero-padding is clearly messed up, the first five observations have been zero-padded and then we have real observations offset by five.

### Versions / Dependencies

Linux Ray 1.7.0

### Reproduction script
Feel free to play with the `USE_CORRECT_SHAPE` flag

```python
import torch
import numpy as np
import gym
from typing import Union, Dict, List, Tuple, Any
import ray
from ray.rllib.models.torch.torch_modelv2 import TorchModelV2
from ray.rllib.utils.typing import ModelConfigDict, TensorType
from ray.rllib.policy.rnn_sequencing import add_time_dimension
from ray.tune import register_env
from ray.rllib.agents.ppo import PPOTrainer
from ray.rllib.examples.env.stateless_cartpole import StatelessCartPole

# Pad to the correct size and crash
# or follow the rnn_sequencing code and don't crash
USE_CORRECT_SHAPE = False

class TestRNN(TorchModelV2, torch.nn.Module):
    def __init__(
        self,
        obs_space: gym.spaces.Space,
        action_space: gym.spaces.Space,
        num_outputs: int,
        model_config: ModelConfigDict,
        name: str,
        **custom_model_kwargs,
    ):
        TorchModelV2.__init__(
            self, obs_space, action_space, num_outputs, model_config, name
        )
        torch.nn.Module.__init__(self)
        self.num_outputs = num_outputs
        self.input_dim = gym.spaces.utils.flatdim(obs_space)
        self.act_space = action_space
        self.act_dim = gym.spaces.utils.flatdim(action_space)
        self.cur_val = None

        self.policy = torch.nn.Linear(self.input_dim, self.act_dim)
        self.vf = torch.nn.Linear(self.input_dim, 1)

    def get_initial_state(self):
        return [torch.zeros(0)]

    def value_function(self):
        assert self.cur_val is not None, "must call forward() first"
        return self.cur_val

    def forward(
        self,
        input_dict: Dict[str, TensorType],
        state: List[TensorType],
        seq_lens: TensorType,
    ) -> Tuple[TensorType, List[TensorType]]:

        flat = input_dict["obs_flat"]

        if USE_CORRECT_SHAPE:
            max_seq_len = seq_lens.max()
        else:
            # max_seq_len here is copied from rllib RNN code
            # see https://github.com/ray-project/ray/blob/2d24ef0d3234867ac329b10ae3a11b9b7119d17b/rllib/models/torch/recurrent_net.py#L75
            # but it doesn't make sense...
            # it should be max_seq_len = seq_len.max()
            max_seq_len = flat.shape[0] // seq_lens.shape[0]

        padded = add_time_dimension(
            flat,
            max_seq_len=max_seq_len,
            framework="torch",
            time_major=False
        )

        B = padded.shape[0]
        T = padded.shape[1]

        # If this fails, then we have "extra" padding in the RNN
        # We shouldn't need to pad the time dimension more than the longest
        # sequence
        if seq_lens.max() != T:
            print(f'seq_lens.max() is {seq_lens.max()} but input temporal dim is {T}')
            print(flat.reshape(B, T, -1) == 0)
            raise Exception('seq_len mismatch')

        flattened = padded.reshape(-1, padded.shape[-1])
        logits = self.policy(flattened)
        self.cur_val = self.vf(flattened).squeeze(1)
        state = state

        return logits, state

register_env(StatelessCartPole.__name__, StatelessCartPole)
MAX_SEQ_LEN = 200
CFG = {
    "env_config": {},
    "framework": "torch",
    "model": {
        "custom_model": TestRNN,
        "max_seq_len": MAX_SEQ_LEN,
    },
    "num_workers": 0,
    "num_gpus": 0,
    "env": StatelessCartPole,
    "horizon": MAX_SEQ_LEN,
}
ray.init(object_store_memory=3e10)
analysis = ray.tune.run(
    PPOTrainer,
    config=CFG,
)
```

### Anything else

Every train step

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

@mvindiola1 with respect to `simple_optimizer`: If I use the following config, I get intermittent crashes at around 50 iters:

```
CFG = {
    "env_config": {},
    "framework": "torch",
    "model": {
        "custom_model": TestRNN,
        "max_seq_len": MAX_SEQ_LEN,
    },
    "num_workers": 0,
    "num_gpus": 0,
    "env": StatelessCartPole,
    "horizon": MAX_SEQ_LEN,
    "simple_optimizer": True,
}
```
This is even if I set `max_seq_len` to the horizon length. Is this the second error condition?


```
2021-11-06 13:33:58,307 ERROR trial_runner.py:924 -- Trial PPO_StatelessCartPole_71918_00000: Error processing event.
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/ray/tune/trial_runner.py", line 890, in _process_trial
    results = self.trial_executor.fetch_result(trial)
  File "/usr/local/lib/python3.8/dist-packages/ray/tune/ray_trial_executor.py", line 788, in fetch_result
    result = ray.get(trial_future[0], timeout=DEFAULT_GET_TIMEOUT)
  File "/usr/local/lib/python3.8/dist-packages/ray/_private/client_mode_hook.py", line 105, in wrapper
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/ray/worker.py", line 1625, in get
    raise value.as_instanceof_cause()
ray.exceptions.RayTaskError(AssertionError): ray::PPO.train_buffered() (pid=45224, ip=172.17.0.2, repr=PPO)
  File "/usr/local/lib/python3.8/dist-packages/ray/tune/trainable.py", line 224, in train_buffered
    result = self.train()
  File "/usr/local/lib/python3.8/dist-packages/ray/rllib/agents/trainer.py", line 682, in train
    raise e
  File "/usr/local/lib/python3.8/dist-packages/ray/rllib/agents/trainer.py", line 668, in train
    result = Trainable.train(self)
  File "/usr/local/lib/python3.8/dist-packages/ray/tune/trainable.py", line 283, in train
    result = self.step()
  File "/usr/local/lib/python3.8/dist-packages/ray/rllib/agents/trainer_template.py", line 206, in step
    step_results = next(self.train_exec_impl)
  File "/usr/local/lib/python3.8/dist-packages/ray/util/iter.py", line 756, in __next__
    return next(self.built_iterator)
  File "/usr/local/lib/python3.8/dist-packages/ray/util/iter.py", line 783, in apply_foreach
    for item in it:
  File "/usr/local/lib/python3.8/dist-packages/ray/util/iter.py", line 783, in apply_foreach
    for item in it:
  File "/usr/local/lib/python3.8/dist-packages/ray/util/iter.py", line 843, in apply_filter
    for item in it:
  File "/usr/local/lib/python3.8/dist-packages/ray/util/iter.py", line 843, in apply_filter
    for item in it:
  File "/usr/local/lib/python3.8/dist-packages/ray/util/iter.py", line 783, in apply_foreach
    for item in it:
  File "/usr/local/lib/python3.8/dist-packages/ray/util/iter.py", line 783, in apply_foreach
    for item in it:
  File "/usr/local/lib/python3.8/dist-packages/ray/util/iter.py", line 791, in apply_foreach
    result = fn(item)
  File "/usr/local/lib/python3.8/dist-packages/ray/rllib/execution/train_ops.py", line 64, in __call__
    learner_info = do_minibatch_sgd(
  File "/usr/local/lib/python3.8/dist-packages/ray/rllib/utils/sgd.py", line 104, in do_minibatch_sgd
    for minibatch in minibatches(batch, sgd_minibatch_size):
  File "/usr/local/lib/python3.8/dist-packages/ray/rllib/utils/sgd.py", line 53, in minibatches
    all_slices = samples._get_slice_indices(sgd_minibatch_size)
  File "/usr/local/lib/python3.8/dist-packages/ray/rllib/utils/annotations.py", line 101, in _ctor
    return obj(*args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/ray/rllib/policy/sample_batch.py", line 904, in _get_slice_indices
    assert np.all(self[SampleBatch.SEQ_LENS] < slice_size), \
AssertionError: ERROR: `slice_size` must be larger than the max. seq-len in the batch!
```

--- a/rllib/BUILD
+++ b/rllib/BUILD
@@ -1222,6 +1222,13 @@ py_test(
 )
 
 py_test(
+    name = "policy/tests/test_rnn_sequencing",
+    tags = ["team:ml", "policy"],
+    size = "small",
+    srcs = ["policy/tests/test_rnn_sequencing.py"]
+)
+
+py_test(
     name = "policy/tests/test_sample_batch",
     tags = ["team:ml", "policy"],
     size = "small",

--- a/rllib/policy/rnn_sequencing.py
+++ b/rllib/policy/rnn_sequencing.py
@@ -134,6 +134,8 @@ def pad_batch_to_sequences_of_same_size(
     for i, k in enumerate(state_keys):
         batch[k] = initial_states[i]
     batch[SampleBatch.SEQ_LENS] = np.array(seq_lens)
+    if dynamic_max:
+        batch.max_seq_len = max(seq_lens)
 
     if log_once("rnn_ma_feed_dict"):
         logger.info("Padded input for RNN/Attn.Nets/MA:\n\n{}\n".format(

--- a/rllib/policy/tests/test_rnn_sequencing.py
+++ b/rllib/policy/tests/test_rnn_sequencing.py
@@ -1,0 +1,90 @@
+import numpy as np
+import unittest
+
+import ray
+from ray.rllib.policy.rnn_sequencing import pad_batch_to_sequences_of_same_size
+from ray.rllib.policy.sample_batch import SampleBatch
+from ray.rllib.policy.view_requirement import ViewRequirement
+from ray.rllib.utils.test_utils import check
+
+
+class TestRNNSequencing(unittest.TestCase):
+    @classmethod
+    def setUpClass(cls) -> None:
+        ray.init()
+
+    @classmethod
+    def tearDownClass(cls) -> None:
+        ray.shutdown()
+
+    def test_pad_batch_dynamic_max(self):
+        """Test pad_batch_to_sequences_of_same_size when dynamic_max = True"""
+        view_requirements = {
+            "state_in_0": ViewRequirement(
+                "state_out_0",
+                shift=[-1],
+                used_for_training=False,
+                used_for_compute_actions=True,
+                batch_repeat_value=1)
+        }
+        max_seq_len = 20
+        num_seqs = np.random.randint(1, 20)
+        seq_lens = np.random.randint(1, max_seq_len, size=(num_seqs))
+        max_len = np.max(seq_lens)
+        sum_seq_lens = np.sum(seq_lens)
+
+        s1 = SampleBatch(
+            {
+                "a": np.arange(sum_seq_lens),
+                "b": np.arange(sum_seq_lens),
+                "seq_lens": seq_lens,
+                "state_in_0": [[0]] * num_seqs
+            },
+            _max_seq_len=max_seq_len)
+
+        pad_batch_to_sequences_of_same_size(
+            s1,
+            max_seq_len=max_seq_len,
+            feature_keys=["a", "b"],
+            view_requirements=view_requirements)
+        check(s1.max_seq_len, max_len)
+        check(s1["a"].shape[0], max_len * num_seqs)
+        check(s1["b"].shape[0], max_len * num_seqs)
+
+    def test_pad_batch_fixed_max(self):
+        """Test pad_batch_to_sequences_of_same_size when dynamic_max = False"""
+        view_requirements = {
+            "state_in_0": ViewRequirement(
+                "state_out_0",
+                shift="-3:-1",
+                used_for_training=False,
+                used_for_compute_actions=True,
+                batch_repeat_value=1)
+        }
+        max_seq_len = 20
+        num_seqs = np.random.randint(1, 20)
+        seq_lens = np.random.randint(1, max_seq_len, size=(num_seqs))
+        sum_seq_lens = np.sum(seq_lens)
+        s1 = SampleBatch(
+            {
+                "a": np.arange(sum_seq_lens),
+                "b": np.arange(sum_seq_lens),
+                "seq_lens": seq_lens,
+                "state_in_0": [[0]] * num_seqs
+            },
+            _max_seq_len=max_seq_len)
+
+        pad_batch_to_sequences_of_same_size(
+            s1,
+            max_seq_len=max_seq_len,
+            feature_keys=["a", "b"],
+            view_requirements=view_requirements)
+        check(s1.max_seq_len, max_seq_len)
+        check(s1["a"].shape[0], max_seq_len * num_seqs)
+        check(s1["b"].shape[0], max_seq_len * num_seqs)
+
+
+if __name__ == "__main__":
+    import pytest
+    import sys
+    sys.exit(pytest.main(["-v", __file__]))

--- a/rllib/utils/sgd.py
+++ b/rllib/utils/sgd.py
@@ -92,13 +92,23 @@ def do_minibatch_sgd(samples, policies, local_worker, num_sgd_iter,
     # no matter the setup (multi-GPU, multi-agent, minibatch SGD,
     # tf vs torch).
     learner_info_builder = LearnerInfoBuilder(num_devices=1)
-    for policy_id in policies.keys():
+    for policy_id, policy in policies.items():
         if policy_id not in samples.policy_batches:
             continue
 
         batch = samples.policy_batches[policy_id]
         for field in standardize_fields:
             batch[field] = standardized(batch[field])
+
+        # Check to make sure that the sgd_minibatch_size is not smaller
+        # than max_seq_len otherwise this will cause indexing errors while
+        # performing sgd when using a RNN or Attention model
+        if policy.is_recurrent() and \
+           policy.config["model"]["max_seq_len"] > sgd_minibatch_size:
+            raise ValueError("`sgd_minibatch_size` ({}) cannot be smaller than"
+                             "`max_seq_len` ({}).".format(
+                                 sgd_minibatch_size,
+                                 policy.config["model"]["max_seq_len"]))
 
         for i in range(num_sgd_iter):
             for minibatch in minibatches(batch, sgd_minibatch_size):

